### DISCUSSION TITLE
Mechanisms to integrate WebGL / WebGPU with Leptos (perhaps Bevy)

### DISCUSSION THREAD

**Discussion by jreuben11** [leptos-rs/leptos#3187] [Q&A] [answered]
Hi,
For WebDev 3D data visualization, the goto tool is the **ThreeJs** framework https://threejs.org/ and the React ecosystem integrates with this using **React-Three-Fiber** https://r3f.docs.pmnd.rs/getting-started/introduction
I really enjoy the pure Rust dev experience that Leptos provides, and I would like to build a weby 3D data visualization application, but am unsure of the recommended architectural approaches to take here to provide such functionality.
I am aware that **Bevy** Rust game engine supports web 3D https://bevyengine.org/examples-webgpu/#3d-rendering but this is a completey seperate ECS runtime - I have no idea of how to integrate that into a Leptos app (ie not from in an iframe).

My question is:
**A)** What is the guidance on Leptos WebGL or wGPU support ? 
and **B)** what is the recommended mechanism to integrate Bevy with Leptos (if possible) ?

**Comment by gbj** [comment, marked as answer]
This kind of integration is outside the scope of Leptos per se. You can do anything you want by using `wasm-bindgen` and `web-sys` to work directly with browser APIs, but I'm not aware of any Leptos-specific library here.
